src/table/selection/use-selection.ts

```diff
@@ -29,92 +29,85 @@ export function useSelection<T>({
+  items.forEach((item, i) => itemIndexesMap.set(getTrackableValue(trackBy, item), i));
+  const getShiftSelectedItems = (item: T): T[] => {
+    const lastClickedItemIndex = lastClickedItem
+      ? itemIndexesMap.get(getTrackableValue(trackBy, lastClickedItem))
+      : undefined;
+    // We use lastClickedItemIndex to determine if filtering/sorting/pagination
+    // made previously selected item invisible, therefore we reset state for shift-select.
+    if (lastClickedItemIndex !== undefined) {
+      const currentItemIndex = itemIndexesMap.get(getTrackableValue(trackBy, item))!;
       const start = Math.min(currentItemIndex, lastClickedItemIndex);
       const end = Math.max(currentItemIndex, lastClickedItemIndex);
-      items.slice(start, end + 1).forEach(item => requestedItems.put(item));
+      return items.slice(start, end + 1);
     }
-    return requestedItems;
+    return [item];
   };
 
-  const deselectItems = (requestedItems: ItemSet<T>) => {
-    const newSelectedItems: Array<T> = [];
-    selectedItems.forEach(selectedItem => {
-      const toUnselect = requestedItems.has(selectedItem);
-      if (!toUnselect || isItemDisabled(selectedItem)) {
-        newSelectedItems.push(selectedItem);
+  // Select items that are not already selected or disabled.
+  const selectItems = (requestedItems: readonly T[]) => {
+    const newSelectedItems = [...selectedItems];
+    requestedItems.forEach(newItem => {
+      if (!isItemSelected(newItem) && !isItemDisabled(newItem)) {
+        newSelectedItems.push(newItem);
       }
     });
     return newSelectedItems;
   };
 
-  const selectItems = (requestedItems: ItemSet<T>) => {
-    const newSelectedItems = [...selectedItems];
-    requestedItems.forEach(newItem => {
-      const { selected, disabled } = getItemState(newItem);
-      if (!selected && !disabled) {
-        newSelectedItems.push(newItem);
+  // Unselect items unless they are disabled.
+  const deselectItems = (requestedItems: readonly T[]) => {
+    const requestedItemsSet = new ItemSet(trackBy, requestedItems);
+    const newSelectedItems: Array<T> = [];
+    selectedItems.forEach(selectedItem => {
+      const shouldUnselect = requestedItemsSet.has(selectedItem);
+      if (!shouldUnselect || isItemDisabled(selectedItem)) {
+        newSelectedItems.push(selectedItem);
       }
     });
     return newSelectedItems;
   };
 
-  const handleToggleItem = (item: T) => () => {
-    const { disabled, selected } = getItemState(item);
-    if (disabled || (selectionType === 'single' && selected)) {
+  const handleToggleAll = () => {
+    const newSelectedItems = allEnabledItemsSelected ? deselectItems(items) : selectItems(items);
+    fireNonCancelableEvent(onSelectionChange, { selectedItems: newSelectedItems });
+  };
+
+  const handleToggleItem = (item: T) => {
+    if (isItemDisabled(item)) {
       return;
     }
-    if (selectionType === 'single') {
+    if (selectionType === 'single' && !isItemSelected(item)) {
```

Comment:
Why was `&& !isItemSelected(item)` added to the condition?

Does `isItemSelected` return the selection state before or after toggling?

Reply:
That is consistent with the previous line `if (disabled || (selectionType === 'single' && selected))`. The reason behind these conditions is to prevent state update is the user clicks on a disabled control or on a radio button which is already selected. The isItemSelected returns state valid for the current render cycle so before toggling.